### PR TITLE
DRAFT: A new plugin hook for dynamic metadata

### PR DESCRIPTION
**Pull request by brandonrobertz**
Note that this is a WORK IN PROGRESS!

This PR adds the following plugin hook:

    get_metadata(
      datasette=self, key=key, database=database, table=table,
      fallback=fallback
    )

This gets called when we're building our metdata for the rest of the system to use. Datasette merges whatever the plugins return with any local metadata (from metadata.yml/yaml/json) allowing for a live-editable dynamic Datasette. __A major design consideration is this: should Datasette perform the metadata merge? Or should Datasette allow plugins to perform any modifications themselves?__

As a security precation, local meta is *not* overwritable by plugin hooks. The workflow for transitioning to live-meta would be to load the plugin with the full metadata.yaml and save. Then remove the parts of the metadata that you want to be able to change from the file.

I have a WIP dynamic configuration plugin here, for reference: https://github.com/next-LI/datasette-live-config/

### CODE REVIEW

**Comment by brandonrobertz** [comment]
Note that if we went with a "update_metadata" hook, the hook signature would look something like this:

```
update_metadata(
  datasette=self, metadata=metadata, key=key, database=database, table=table,
  fallback=fallback
)
```

And the Datasette function `_metadata_recursive_update(self, orig, updated)` would disappear into the plugins.

I have a WIP dynamic configuration plugin here, for reference: https://github.com/next-LI/datasette-live-config/

**Comment by simonw** [comment]
A few tests failed - Black, the test that checks the docs mention the new hook - th most interesting failing test looks like this one:
```
            updated_metadata["databases"]["fixtures"]["queries"]["magic_parameters"][
                "allow"
            ] = (allow if "query" in permissions else deny)
>           cascade_app_client.ds._metadata = updated_metadata
E           AttributeError: can't set attribute
```

File: datasette/app.py [hunk 1]
```diff
@@ -415,7 +433,24 @@ def metadata(self, key=None, database=None, table=None, fallback=True):
         assert not (
             database is None and table is not None
         ), "Cannot call metadata() with table= specified but not database="
-        databases = self._metadata.get("databases") or {}
+        metadata = {}
+
+        # TODO: change this to update_metadata, pass meta back allow the
+        # plugins to mutate it?
```
Comment (simonw) [inline datasette/app.py:439]: Let's ditch the TODO comments before we merge the PR - they can live in PR comments the issue conversation instead.

File: datasette/app.py [hunk 2]
```diff
@@ -424,7 +459,10 @@ def metadata(self, key=None, database=None, table=None, fallback=True):
                 table
             ) or {}
             search_list.insert(0, table_metadata)
-        search_list.append(self._metadata)
+
+        # TODO: get full data for here, we need to work down to tables from that
+        # as opposed to simply returning table data above?
```
Comment (simonw) [inline datasette/app.py:464]: I don't fully understand this comment

File: datasette/app.py [hunk 3]
```diff
@@ -440,6 +478,10 @@ def metadata(self, key=None, database=None, table=None, fallback=True):
                 m.update(item)
             return m
 
+    @property
+    def _metadata(self):
+        return self.metadata()
+
```
Comment (simonw) [inline datasette/app.py:484]: If this is for backwards compatibility I don't think it's necessary - `_metadata` has never been part of Datasette's public documented API.

**Comment by brandonrobertz** [comment]
If this is a concept ACK then I will move onto fixing the tests (adding new ones) and updating the documentation for the new plugin hook.

**Comment by simonw** [comment]
Yes, let's move ahead with getting this into an alpha.

**Comment by simonw** [comment]
The only test failure is Black. I'm going to merge this and then reformat.

**Pull request merged**